enhancements/local-storage/subprovisioner-integration-into-lvms.md

```diff
@@ -0,0 +1,255 @@
+  
+  const (
+    DeviceAccessPolicyShared DeviceAccessPolicy = "shared"
+    DeviceAccessPolicyNodeLocal  DeviceAccessPolicy = "nodeLocal"
+  )
+
+  // LVMClusterSpec defines the desired state of LVMCluster
+  type LVMClusterSpec struct {
+    // Important: Run "make" to regenerate code after modifying this file
+    
+    // Tolerations to apply to nodes to act on
+    // +optional
+    Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+    // Storage describes the deviceClass configuration for local storage devices
+    // +Optional
+    Storage Storage `json:"storage,omitempty"`
+  }
+  type Storage struct {
+    // DeviceClasses are a rules that assign local storage devices to volumegroups that are used for creating lvm based PVs
+    // +Optional
+    DeviceClasses []DeviceClass `json:"deviceClasses,omitempty"`
+  }
+  
+  type DeviceClass struct {
+    // Name of the class, the VG and possibly the storageclass.
+    // Validations to confirm that this field can be used as metadata.name field in storageclass
+    // ref: https://github.com/kubernetes/apimachinery/blob/de7147/pkg/util/validation/validation.go#L209
+    // +kubebuilder:validation:MaxLength=245
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:Pattern="^[a-z0-9]([-a-z0-9]*[a-z0-9])?$"
+    Name string `json:"name,omitempty"`
+    
+    // DeviceSelector is a set of rules that should match for a device to be included in the LVMCluster
+    // +optional
+    DeviceSelector *DeviceSelector `json:"deviceSelector,omitempty"`
+    
+    // NodeSelector chooses nodes on which to create the deviceclass
+    // +optional
+    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+    
+    // ThinPoolConfig contains configurations for the thin-pool
+    // +optional
+    ThinPoolConfig *ThinPoolConfig `json:"thinPoolConfig,omitempty"`
+    
+    // Default is a flag to indicate whether the device-class is the default.
+    // This will mark the storageClass as default.
+    // +optional
+    Default bool `json:"default,omitempty"`
+    
+    // FilesystemType sets the filesystem the device should use.
+    // For shared deviceClasses, this field must be set to "" or none.
+    // +kubebuilder:validation:Enum=xfs;ext4;none;""
+    // +kubebuilder:default=xfs
+    // +optional
+    FilesystemType DeviceFilesystemType `json:"fstype,omitempty"`
+    
+    // Policy defines the policy for the deviceClass.
+    // TECH PREVIEW: shared will allow accessing the deviceClass from multiple nodes.
+    // The deviceClass will then be configured via shared volume group.
+    // +optional	  
+    // +kubebuilder:validation:Enum=shared;local
++   DeviceAccessPolicy DeviceAccessPolicy `json:"deviceAccessPolicy,omitempty"`
+  }
+
+  type ThinPoolConfig struct {
+    // Name of the thin pool to be created. Will only be used for node-local storage, 
+    // since shared volume groups will create a thin pool with the same name as the volume group.
```

Comment:
Subprovisioner creates its own thin pools within the shared volume group, but it is more along the lines of one thin pool per PVC rather than one for the entire volume group.  Is this comment trying to give too much internal detail about what subprovisioner will do, in relation to the higher-level notion that a client of LVMS merely needs to choose between local (existing TopoLVM) and shared (new subprovisioner) implementations?

Reply:
I think this would eventually also result in https://github.com/openshift/enhancements/pull/1620#discussion_r1593038540